---
title: "Processing MaxQuant output for PGNEXUS proteome data"
author: "Angel Liang"
date: "22/02/2020"
output: html_document
---


# Set the running environment

## Packages and directories

```{r}

library(tidyverse)
# library(gtools)
# library(extrafont)
# loadfonts(device="win")
windowsFonts("Helvetica" = windowsFont("Helvetica"))
library(RColorBrewer)

library(ggplot2)
# library(kohonen)
# library(genefilter)
# library(gplots)
# library(lattice)
# library(svglite)
# library(scales)

# # library(biomaRt)
# library(systemPipeR)
# library(GOstats)
# library(PFAM.db)
# library(bc3net)
# 
# library(ggdendro)
library(data.table)
# library(Rfast)


maxquant_sp.canon_sp.isoform_junctions.3FT_results_dir <- "Z:/PGNEXUS_kassem_MSC/Kassem_OB_fastqc/proteome_validation/results_maxquant/swissprot.hsa.canonical.isoforms_junctions.3FT.som.qvalue0.01.dpsi0.2/txt/"

# maxquant_junctions.3FT_results_dir <- "Z:/PGNEXUS_kassem_MSC/Kassem_OB_fastqc/proteome_validation/results_maxquant/run_1_PGNEXUS_OBseries_allsamples_angel.junctions.3FT/txt/"

R_processing_results <- paste("Z:/PGNEXUS_kassem_MSC/Kassem_OB_fastqc/proteome_validation/results_maxquant/R_processing_results/", sep="")

if(! dir.exists(R_processing_results) ) {
     dir.create(R_processing_results, recursive = TRUE)}


```

# read the tables into environment

```{r}

list_maxquant_run_dirs <- list("PGNEXUS_OBseries_sp.canon_sp.isoform_junctions.3FT" = maxquant_sp.canon_sp.isoform_junctions.3FT_results_dir)
list_output.table_categories <- list("msms.txt", "evidence.txt", "proteinGroups.txt")

list_for_maxquant_table_import <- purrr::cross2(list_maxquant_run_dirs, list_output.table_categories)

list_imported_maxquant_tibbles <- purrr::map(.x = list_for_maxquant_table_import, .f = ~read.delim(paste(.x[[1]], .x[[2]], sep = ""), stringsAsFactors = FALSE, row.names = NULL, sep = "\t") %>% as_tibble)

```

# global processing of tables

```{r}

# at this stage, evidence.txt is in every 1, 3, 5, 7 etc... elements
# msms.txt is 

vector_evidence.txt_list.elements <- c(1:(length(list_imported_maxquant_tibbles) / 2))
list_imported_maxquant_tibbles_processed <- list_imported_maxquant_tibbles %>% modify_at(.at = vector_evidence.txt_list.elements, 
                                                                                       .f = ~dplyr::filter(.x, MS.MS.count != 0) %>% dplyr::arrange(., PEP))

```

# Data triaging

## plot the PEP and score distributions of CON, SP and 3FT identified peptides

### rearrange the evidence.txt tables for ggplot

```{r}

# subset evidence tables
list_evidence_tibbles <- list_imported_maxquant_tibbles_processed[vector_evidence.txt_list.elements]

# append database run columns then peptide class
list_dbrun.info <- names(list_maxquant_run_dirs)

# function to create peptide class column
peptide_class_column <- function(evidence_tibble) {
  
  col <- rep(NA, times = nrow(evidence_tibble))
  
  # record the columns of contaminants, reverse sequences, swissprot matches, junction and exon matches.
  cons <- grep(x = evidence_tibble$Proteins, pattern = "(^CON_|;CON_)", ignore.case = FALSE)
  revs <- grep(x = evidence_tibble$Leading.proteins, pattern = "^REV_", ignore.case = FALSE)
  sps <- grep(x = evidence_tibble$Proteins, pattern = "(^[A-Z][0-9].*|;[A-Z][0-9].*)", ignore.case = FALSE)
  junctions <- grep(x = evidence_tibble$Proteins, pattern = "Junction_.*", ignore.case = FALSE)
  exons <- grep(x = evidence_tibble$Proteins, pattern = "Exon_.*", ignore.case = FALSE)
  
  col[cons] <- "contaminants"
  col[revs] <- "reversed_sequences"
  col[sps] <- "swissprot_hits"
  col[junctions] <- "junction_3FT_hits"
  col[exons] <- "exon_3FT_hits"
  
  return(col)
  
}

list_evidence_tibbles <- purrr::map2(.x = list_evidence_tibbles, .y = list_dbrun.info, .f = ~add_column(.x, "dbrun" = .y) %>% add_column("peptide_class" = peptide_class_column(.x)))

# melt into long tables for ggplot faceting by peptide class and database run
long_tibble_evidence_tibbles_facet.peptideclass.dbrun <- list_evidence_tibbles %>% rbindlist(fill = TRUE) %>% as_tibble

```

### ggplot of number of hits in each category

```{r}

ggplot(long_tibble_evidence_tibbles_facet.peptideclass.dbrun, aes(x = peptide_class, fill = PEP <= 0.01)) +
  geom_bar() +
  facet_wrap(~dbrun, scales = "free") +
  ggtitle(paste("Bar plot of number of hits in each peptide category", sep = "")) +
  xlab("Peptide category") +
  scale_x_discrete(limits = c("swissprot_hits", "junction_3FT_hits", "contaminants", "reversed_sequences"), labels = c("Swissprot hits", "Junction from 3FT", "Contaminant", "Reversed sequence")) + 
  ylab(expression(Number~of~total~hits)) +
  scale_fill_manual(limits = c("TRUE", "FALSE"), labels = c("PEP < 0.01", "PEP > 0.01"), values = c("blue2", "grey75"), name = "Hit significance") +
  theme_bw() +
  theme(text = element_text(family = "Helvetica")) +
  ggsave(filename = paste(R_processing_results, "qualityplot_number.of.hits_per_peptide.class", ".pdf", sep = ""), device = "pdf", dpi = 600, width = 45, height = 25, units = "cm") +
  ggsave(filename = paste(R_processing_results, "qualityplot_number.of.hits_per_peptide.class", ".svg", sep = ""), device = "svg", dpi = 600, width = 45, height = 25, units = "cm")

ggplot(long_tibble_evidence_tibbles_facet.peptideclass.dbrun, aes(x = peptide_class, fill = PEP <= 0.01)) +
  geom_bar(aes(y = log10(..count..))) +
  facet_wrap(~dbrun, scales = "free") +
  ggtitle(paste("Bar plot of number of hits in each peptide category", sep = "")) +
  xlab("Peptide category") +
  scale_x_discrete(limits = c("swissprot_hits", "junction_3FT_hits", "contaminants", "reversed_sequences"), labels = c("Swissprot hits", "Junction from 3FT", "Contaminant", "Reversed sequence")) + 
  ylab(expression(Number~of~total~hits~(log[10]))) +
  # scale_y_continuous(trans = "log10") +
  scale_fill_manual(limits = c("TRUE", "FALSE"), labels = c("PEP < 0.01", "PEP > 0.01"), values = c("blue2", "grey75"), name = "Hit significance") +
  theme_bw() +
  theme(text = element_text(family = "Helvetica")) +
  ggsave(filename = paste(R_processing_results, "qualityplot_number.of.hits_per_peptide.class_log10", ".pdf", sep = ""), device = "pdf", dpi = 600, width = 45, height = 25, units = "cm") +
  ggsave(filename = paste(R_processing_results, "qualityplot_number.of.hits_per_peptide.class_log10", ".svg", sep = ""), device = "svg", dpi = 600, width = 45, height = 25, units = "cm")

```

### ggplot of PEP distributions below 0.05

```{r}

# line - absolute counts
ggplot(long_tibble_evidence_tibbles_facet.peptideclass.dbrun, aes(x = PEP, color = peptide_class)) +
  geom_density(aes(y = ..count..)) +
  # facet_wrap(~dbrun + peptide_class, scales = "free") +
  ggtitle(expression(Comparison~of~peptide~p-value~(PEP)~distributions~(PEP <= 0.05~only))) +
  xlab(expression(PEP)) +
  # xlim(c(0, 1)) +
  scale_x_continuous(limits = c(-0.005, 0.051), breaks = c(-0.005, seq(0, 0.051, 0.005))) +
  ylab("Frequency (smoothed)") +
  scale_color_manual(limits = c("swissprot_hits", "junction_3FT_hits", "contaminants", "reversed_sequences"), labels = c("Swissprot hits", "Junction from 3FT", "Contaminant", "Reversed sequence"), values = c("red", "blue2", "grey75", "lawngreen"), name = "Peptide Hit Type") + 
  theme_bw() +
  theme(text = element_text(family = "Helvetica")) +
  ggsave(filename = paste(R_processing_results, "qualityplot_peptide.PEP.less.than.0.05_distribution_per_peptide.class_line.absolute", ".pdf", sep = ""), device = "pdf", dpi = 600, width = 45, height = 25, units = "cm") +
  ggsave(filename = paste(R_processing_results, "qualityplot_peptide.PEP.less.than.0.05_distribution_per_peptide.class_line.absolute", ".svg", sep = ""), device = "svg", dpi = 600, width = 45, height = 25, units = "cm")

# line - normalised counts
ggplot(long_tibble_evidence_tibbles_facet.peptideclass.dbrun, aes(x = PEP, color = peptide_class)) +
  geom_density(aes(y = ..density..)) +
  # facet_wrap(~dbrun + peptide_class, scales = "free") +
  ggtitle(expression(Comparison~of~peptide~p-value~(PEP)~distributions~(PEP <= 0.05~only))) +
  xlab(expression(PEP)) +
  # xlim(c(0, 1)) +
  scale_x_continuous(limits = c(-0.005, 0.051), breaks = c(-0.005, seq(0, 0.051, 0.005))) +
  ylab("Frequency (smoothed)") +
  scale_color_manual(limits = c("swissprot_hits", "junction_3FT_hits", "contaminants", "reversed_sequences"), labels = c("Swissprot hits", "Junction from 3FT", "Contaminant", "Reversed sequence"), values = c("red", "blue2", "grey75", "lawngreen"), name = "Peptide Hit Type") + 
  theme_bw() +
  theme(text = element_text(family = "Helvetica")) +
  ggsave(filename = paste(R_processing_results, "qualityplot_peptide.PEP.less.than.0.05_distribution_per_peptide.class_line.normalised", ".pdf", sep = ""), device = "pdf", dpi = 600, width = 45, height = 25, units = "cm") +
  ggsave(filename = paste(R_processing_results, "qualityplot_peptide.PEP.less.than.0.05_distribution_per_peptide.class_line.normalised", ".svg", sep = ""), device = "svg", dpi = 600, width = 45, height = 25, units = "cm")

# bar
ggplot(long_tibble_evidence_tibbles_facet.peptideclass.dbrun, aes(x = PEP, fill = peptide_class)) +
  geom_histogram(binwidth = 0.0005, aes(y = ..count..)) +
  facet_wrap(~dbrun + peptide_class, scales = "free") +
  ggtitle(expression(Comparison~of~peptide~p-value~(PEP)~distributions~(PEP <= 0.05~only))) +
  xlab(expression(PEP)) +
  # xlim(c(0, 1)) +
  scale_x_continuous(limits = c(-0.001, 0.051), breaks = c(-0.001, seq(0, 0.051, 0.005))) +
  ylab("Frequency (smoothed)") +
  scale_fill_manual(limits = c("swissprot_hits", "junction_3FT_hits", "contaminants", "reversed_sequences"), labels = c("Swissprot hits", "Junction from 3FT", "Contaminant", "Reversed sequence"), values = c("red", "blue2", "grey75", "lawngreen"), name = "Peptide Hit Type") + 
  theme_bw() +
  theme(text = element_text(family = "Helvetica")) +
  ggsave(filename = paste(R_processing_results, "qualityplot_peptide.PEP.less.than.0.05_distribution_per_peptide.class_bar", ".pdf", sep = ""), device = "pdf", dpi = 600, width = 45, height = 25, units = "cm") +
  ggsave(filename = paste(R_processing_results, "qualityplot_peptide.PEP.less.than.0.05_distribution_per_peptide.class_bar", ".svg", sep = ""), device = "svg", dpi = 600, width = 45, height = 25, units = "cm")

```

### ggplot of score distributions

```{r}

# absolute counts
ggplot(long_tibble_evidence_tibbles_facet.peptideclass.dbrun, aes(x = Score + 1, colour = peptide_class)) +
  geom_density(aes(y = ..count.. + 1)) +
  facet_wrap(~dbrun, scales = "free") +
  ggtitle(paste("Comparison of peptide score distributions", sep = "")) +
  xlab(expression(Score[MaxQuant] + 1)) +
  scale_x_continuous(trans = "log10") +
  ylab("Frequency (absolute)") +
  # scale_y_continuous(trans = "log10") +
  scale_color_manual(limits = c("swissprot_hits", "junction_3FT_hits", "contaminants", "reversed_sequences"), labels = c("Swissprot hits", "Junction from 3FT", "Contaminant", "Reversed sequence"), values = c("red", "blue2", "grey75", "lawngreen"), name = "Peptide Hit Type") + 
  theme_bw() +
  theme(text = element_text(family = "Helvetica")) +
  ggsave(filename = paste(R_processing_results, "qualityplot_peptide.score_distribution_per_peptide.class_absolute.counts", ".pdf", sep = ""), device = "pdf", dpi = 600, width = 45, height = 25, units = "cm") +
  ggsave(filename = paste(R_processing_results, "qualityplot_peptide.score_distribution_per_peptide.class_absolute.counts", ".svg", sep = ""), device = "svg", dpi = 600, width = 45, height = 25, units = "cm")

# normalised counts
ggplot(long_tibble_evidence_tibbles_facet.peptideclass.dbrun, aes(x = Score + 1, colour = peptide_class)) +
  geom_density(aes(y = ..density.. + 1)) +
  facet_wrap(~dbrun, scales = "free") +
  ggtitle(paste("Comparison of peptide score distributions", sep = "")) +
  xlab(expression(Score[MaxQuant] + 1)) +
  scale_x_continuous(trans = "log10") +
  ylab("Frequency (normalised)") +
  # scale_y_continuous(trans = "log10") +
  scale_color_manual(limits = c("swissprot_hits", "junction_3FT_hits", "contaminants", "reversed_sequences"), labels = c("Swissprot hits", "Junction from 3FT", "Contaminant", "Reversed sequence"), values = c("red", "blue2", "grey75", "lawngreen"), name = "Peptide Hit Type") + 
  theme_bw() +
  theme(text = element_text(family = "Helvetica")) +
  ggsave(filename = paste(R_processing_results, "qualityplot_peptide.score_distribution_per_peptide.class_normalised.counts", ".pdf", sep = ""), device = "pdf", dpi = 600, width = 45, height = 25, units = "cm") +
  ggsave(filename = paste(R_processing_results, "qualityplot_peptide.score_distribution_per_peptide.class_normalised.counts", ".svg", sep = ""), device = "svg", dpi = 600, width = 45, height = 25, units = "cm")

```

### plot of score vs. PEP

```{r}

ggplot(long_tibble_evidence_tibbles_facet.peptideclass.dbrun %>% dplyr::distinct(., PEP, Score, .keep_all = TRUE) %>% dplyr::filter(., PEP <= 1), aes(x = log10(PEP + 1), y = Score + 1, colour = peptide_class)) +
  geom_point() +
  geom_smooth(formula = y ~ x) +
  facet_wrap(~dbrun, scales = "free") +
  ggtitle(expression(Correlation~plot~of~peptide~score~vs.~PEP~(PEP <= 0.05~only))) +
  xlab(expression(log[10](PEP[MaxQuant] + 1))) +
  ylab(expression(Score[MaxQuant])) +
  scale_y_continuous(trans = "log10", breaks = seq(0, 500, 50)) +
  theme_bw() +
  theme(text = element_text(family = "Helvetica")) +
  ggsave(filename = paste(R_processing_results, "qualityplot_correlation_score_vs_PEP", ".pdf", sep = ""), device = "pdf", dpi = 600, width = 45, height = 25, units = "cm") +
  ggsave(filename = paste(R_processing_results, "qualityplot_correlation_score_vs_PEP", ".svg", sep = ""), device = "svg", dpi = 600, width = 45, height = 25, units = "cm")

```

## evaluate the local FDR of peptides only

```{r}

# get the indices of the tables in list which contain the combined sp + junction results
sp.junctions.combined_index <- c(1)



```

