---
title: "Sorting JUM output files and analysing them"
author: "Angel Liang"
date: "10/21/2019"
output: pdf_document
---

## Set the running environment

```{r}

library(tidyverse)
library(dplyr)
library(purrr)
library(data.table)

results_dir <- "C:/Users/angel/Documents/run_4_BM_alltimeseries_pvalue1/results/"

```

## read all the final tables into environment

```{r}

list_of_timepoint_comparisons_final <- read.delim(paste(results_dir, "list_of_timepoint_comparisons_final.txt", sep = ""), stringsAsFactors = FALSE, sep = "\t", header = FALSE, row.names = NULL)

print(head(list_of_timepoint_comparisons_final))

for (comparison in list_of_timepoint_comparisons_final[, 1])
{
  
  for (AS_event in c("A3SS_events", "A5SS_events", "cassette_exon_events", "composite_events", "intron_retention", "MXE_events"))
  {
    
    assign(x = paste(comparison, "_", AS_event, "_simplified", sep = ""), value = read.delim(file = paste(results_dir, "final_JUM_output_", comparison, "/", list.files(path = paste(results_dir, "final_JUM_output_", comparison, "/", sep = ""), pattern = paste("(.)", AS_event, "(.*)simplified.txt", sep = "")), sep = ""), sep = "\t", header = TRUE, stringsAsFactors = FALSE, row.names = NULL, na.strings = c("NONE", "NA", "INF", "Inf")))
  
  }
  
}

# this must be equal to the total number of comparisons

length(ls(pattern = "(.*)simplified")) / 6

```

## chopping files into a net summary table

note: the final simplified intron retention table outputted by JUM has two header rows. Remove those.

### A3SS

```{r}

list_of_A3SS_simplified_output_tables <- ls(pattern = "(.*)(_)(A3SS_events)(_simplified$)")

print(paste("There are", length(list_of_A3SS_simplified_output_tables), "output tables loaded for A3SS events:", sep = " "))  

print(list_of_A3SS_simplified_output_tables)

for (simplified_table_name in as.list(list_of_A3SS_simplified_output_tables))
{
  
  simplified_table <- get(simplified_table_name)

  simplified_table <- simplified_table[, c(5, 1:2, 6:9)]

  # annotating the column names to be comparison name specific
  colnames(simplified_table)[5] <- gsub(x = simplified_table_name, pattern = "(.*)_A3SS_events_simplified", replacement = "pvalue_\\1")
  colnames(simplified_table)[6] <- gsub(x = simplified_table_name, pattern = "(.*)_A3SS_events_simplified", replacement = "qvalue_\\1")
  
  
  assign(x = paste(simplified_table_name, "_shuffled", sep = ""), value = simplified_table)
  
}

list_of_A3SS_simplified_output_tables_shuffled <- ls(pattern = "(.*)(_)(A3SS_events)(_simplified_shuffled$)")

print(paste("Shuffled", length(list_of_A3SS_simplified_output_tables_shuffled), "output tables for A3SS events:", sep = " "))  

print(list_of_A3SS_simplified_output_tables_shuffled)

# conduct full join of all the shuffled tables

A3SS_summary_table <- purrr::map(list_of_A3SS_simplified_output_tables_shuffled, get) %>% reduce(full_join, by = c("common_5_SS_coor", "Gene", "AS_event_ID", "A3SS_coordinates"))

print(dim(A3SS_summary_table))


```

### A5SS

```{r}

list_of_A5SS_simplified_output_tables <- ls(pattern = "(.*)(_)(A5SS_events)(_simplified$)")

print(paste("There are", length(list_of_A5SS_simplified_output_tables), "output tables loaded for A5SS events:", sep = " "))  

print(list_of_A5SS_simplified_output_tables)

for (simplified_table_name in as.list(list_of_A5SS_simplified_output_tables))
{
  
  simplified_table <- get(simplified_table_name)

  simplified_table <- simplified_table[, c(5, 1:2, 6:9)]

  # annotating the column names to be comparison name specific
  colnames(simplified_table)[5] <- gsub(x = simplified_table_name, pattern = "(.*)_A5SS_events_simplified", replacement = "pvalue_\\1")
  colnames(simplified_table)[6] <- gsub(x = simplified_table_name, pattern = "(.*)_A5SS_events_simplified", replacement = "qvalue_\\1")
  
  
  assign(x = paste(simplified_table_name, "_shuffled", sep = ""), value = simplified_table)
  
}

list_of_A5SS_simplified_output_tables_shuffled <- ls(pattern = "(.*)(_)(A5SS_events)(_simplified_shuffled$)")

print(paste("Shuffled", length(list_of_A5SS_simplified_output_tables_shuffled), "output tables for A5SS events:", sep = " "))  

print(list_of_A5SS_simplified_output_tables_shuffled)

# conduct full join of all the shuffled tables

A5SS_summary_table <- purrr::map(list_of_A5SS_simplified_output_tables_shuffled, get) %>% reduce(full_join, by = c("common_3_SS_coor", "Gene", "AS_event_ID", "A5SS_coordinates"))

print(dim(A5SS_summary_table))


```

### cassette exon

```{r}

list_of_cassette_exon_simplified_output_tables <- ls(pattern = "(.*)(_)(cassette_exon_events)(_simplified$)")

print(paste("There are", length(list_of_cassette_exon_simplified_output_tables), "output tables loaded for cassette exon events:", sep = " "))  

print(list_of_cassette_exon_simplified_output_tables)

for (simplified_table_name in as.list(list_of_cassette_exon_simplified_output_tables))
{
  
  simplified_table <- get(simplified_table_name)

  simplified_table <- simplified_table[, c(5, 8, 1:2, 6:7, 9:11)]

  # annotating the column names to be comparison name specific
  colnames(simplified_table)[7] <- gsub(x = simplified_table_name, pattern = "(.*)_cassette_exon_events_simplified", replacement = "pvalue_\\1")
  colnames(simplified_table)[8] <- gsub(x = simplified_table_name, pattern = "(.*)_cassette_exon_events_simplified", replacement = "qvalue_\\1")
  
  
  assign(x = paste(simplified_table_name, "_shuffled", sep = ""), value = simplified_table)
  
}

list_of_cassette_exon_simplified_output_tables_shuffled <- ls(pattern = "(.*)(_)(cassette_exon_events)(_simplified_shuffled$)")

print(paste("Shuffled", length(list_of_cassette_exon_simplified_output_tables_shuffled), "output tables for cassette exon events:", sep = " "))  

print(list_of_cassette_exon_simplified_output_tables_shuffled)

# conduct full join of all the shuffled tables

cassette_exon_summary_table <- purrr::map(list_of_cassette_exon_simplified_output_tables_shuffled, get) %>% reduce(full_join, by = c("upstream_exon_end_coor", "downstream_exon_start_coor", "Gene", "AS_event_ID", "cassette_exon_start_coor", "cassette_exon_end_coor"))

print(dim(cassette_exon_summary_table))


```



### MXE events

```{r}

list_of_MXE_simplified_output_tables <- ls(pattern = "(.*)(_)(MXE_events)(_simplified$)")

print(paste("There are", length(list_of_MXE_simplified_output_tables), "output tables loaded for MXE events:", sep = " "))  

print(list_of_MXE_simplified_output_tables)

for (simplified_table_name in as.list(list_of_MXE_simplified_output_tables))
{
  
  simplified_table <- get(simplified_table_name)

  simplified_table <- simplified_table[, c(5, 7, 1:2, 6, 8:10)]

  # annotating the column names to be comparison name specific
  colnames(simplified_table)[6] <- gsub(x = simplified_table_name, pattern = "(.*)_MXE_events_simplified", replacement = "pvalue_\\1")
  colnames(simplified_table)[7] <- gsub(x = simplified_table_name, pattern = "(.*)_MXE_events_simplified", replacement = "qvalue_\\1")
  
  
  assign(x = paste(simplified_table_name, "_shuffled", sep = ""), value = simplified_table)
  
}

list_of_MXE_simplified_output_tables_shuffled <- ls(pattern = "(.*)(_)(MXE_events)(_simplified_shuffled$)")

print(paste("Shuffled", length(list_of_MXE_simplified_output_tables_shuffled), "output tables for MXE events:", sep = " "))  

print(list_of_MXE_simplified_output_tables_shuffled)

# conduct full join of all the shuffled tables

MXE_summary_table <- purrr::map(list_of_MXE_simplified_output_tables_shuffled, get) %>% reduce(full_join, by = c("upstream_exon_end_coor", "downstream_exon_start_coor", "Gene", "AS_event_ID", "MXE_exon_coordinates"))

print(dim(MXE_summary_table))


```

### composite events

```{r}

list_of_composite_events_simplified_output_tables <- ls(pattern = "(.*)(_)(composite_events)(_simplified$)")

print(paste("There are", length(list_of_composite_events_simplified_output_tables), "output tables loaded for composite events:", sep = " "))  

print(list_of_composite_events_simplified_output_tables)

for (simplified_table_name in as.list(list_of_composite_events_simplified_output_tables))
{
  
  simplified_table <- get(simplified_table_name)

  simplified_table <- simplified_table[, c(5, 1:2, 6:8)]

  # annotating the column names to be comparison name specific
  colnames(simplified_table)[4] <- gsub(x = simplified_table_name, pattern = "(.*)_composite_events_simplified", replacement = "pvalue_\\1")
  colnames(simplified_table)[5] <- gsub(x = simplified_table_name, pattern = "(.*)_composite_events_simplified", replacement = "qvalue_\\1")
  
  
  assign(x = paste(simplified_table_name, "_shuffled", sep = ""), value = simplified_table)
  
}

list_of_composite_events_simplified_output_tables_shuffled <- ls(pattern = "(.*)(_)(composite_events)(_simplified_shuffled$)")

print(paste("Shuffled", length(list_of_composite_events_simplified_output_tables_shuffled), "output tables for composite events:", sep = " "))  

print(list_of_composite_events_simplified_output_tables_shuffled)

# conduct full join of all the shuffled tables

composite_events_summary_table <- purrr::map(list_of_composite_events_simplified_output_tables_shuffled, get) %>% reduce(full_join, by = c("Composite_coordinates", "Gene", "AS_event_ID"))

print(dim(composite_events_summary_table))


```

### IR events

```{r}

list_of_IR_events_simplified_output_tables <- ls(pattern = "(.*)(_)(intron_retention)(_simplified$)")

print(paste("There are", length(list_of_IR_events_simplified_output_tables), "output tables loaded for IR events:", sep = " "))  

print(list_of_IR_events_simplified_output_tables)

for (simplified_table_name in as.list(list_of_IR_events_simplified_output_tables))
{
  
  simplified_table <- get(simplified_table_name)

  simplified_table <- simplified_table[(simplified_table$chromosome != "chromosome") & !(is.na(simplified_table$Gene)) & !(is.na(simplified_table$pvalue)), ]
  
  simplified_table <- simplified_table[, c(5, 6, 1:2, 7:9)]

  # annotating the column names to be comparison name specific
  colnames(simplified_table)[5] <- gsub(x = simplified_table_name, pattern = "(.*)_intron_retention_simplified", replacement = "pvalue_\\1")
  colnames(simplified_table)[6] <- gsub(x = simplified_table_name, pattern = "(.*)_intron_retention_simplified", replacement = "qvalue_\\1")
  
  
  assign(x = paste(simplified_table_name, "_shuffled", sep = ""), value = simplified_table)
  
}

list_of_IR_events_simplified_output_tables_shuffled <- ls(pattern = "(.*)(_)(intron_retention)(_simplified_shuffled$)")

print(paste("Shuffled", length(list_of_IR_events_simplified_output_tables_shuffled), "output tables for IR events:", sep = " "))  

print(list_of_IR_events_simplified_output_tables_shuffled)

# conduct full join of all the shuffled tables

IR_events_summary_table <- purrr::map(list_of_IR_events_simplified_output_tables_shuffled, get) %>% reduce(full_join, by = c("retained_intron_start", "retained_intron_end", "Gene", "AS_event_ID"))

print(dim(IR_events_summary_table))


```

# discard the superfluous columns for each table containing each splice mode

#### simplified_table <- simplified_table[(simplified_table$chromosome != "chromosome") & !(is.na(simplified_table$Gene)) & !(is.na(simplified_table$pvalue)), ]


```{r}

# create column containing every simplified output table, subset only the rows of relevance and create summary tables with a full join
list_of_A3SS_simplified_output_table_names <- ls(pattern = "(.*)(_)(A3SS_events)(_simplified$)")
print(paste("There are", length(list_of_A3SS_simplified_output_table_names), "output tables loaded for A3SS events:", sep = " "))  
## convert from column of names to column of dfs
list_of_A3SS_simplified_output_tables <- purrr::map(list_of_A3SS_simplified_output_table_names, get)
## subset
list_of_A3SS_simplified_output_tables_simplified <- purrr::map(list_of_A3SS_simplified_output_tables, ~select(.x, 1, 2, 7, 8, 9))



list_of_A5SS_simplified_output_table_names <- ls(pattern = "(.*)(_)(A5SS_events)(_simplified_shuffled$)")
list_of_cassette_exon_simplified_output_table_names <- ls(pattern = "(.*)(_)(cassette_exon_events)(_simplified$)")
list_of_MXE_simplified_output_table_names <- ls(pattern = "(.*)(_)(MXE_events)(_simplified$)")
list_of_composite_events_simplified_output_table_names <- ls(pattern = "(.*)(_)(composite_events)(_simplified$)")
list_of_IR_events_simplified_output_table_names <- ls(pattern = "(.*)(_)(intron_retention)(_simplified$)")

column_of_simplified_output_table_names <- c("list_of_A3SS_simplified_output_tables", "list_of_A5SS_simplified_output_tables", "list_of_cassette_exon_simplified_output_tables", "list_of_MXE_simplified_output_tables", "list_of_composite_events_simplified_output_tables", "list_of_IR_events_simplified_output_tables")

column_of_simplified_output_tables <- 


column_of_simplified_output_tables <- purrr::map(column_of_simplified_output_table_names, get)

purrr::map(column_of_simplified_output_tables, ~select(x., ))

purrr::map2()

```


### append the name of the splice mode to each data frame 



```{r}

column_of_summary_tables_per_splicemode <- purrr::map(ls(pattern = "(.*)_summary_table$"), get)

column_of_splicemode_names <- ls(pattern = "(.*)_summary_table$") 

column_of_splicemode_names <- gsub(column_of_splicemode_names, pattern = "(.*)_summary_table$", replacement = "\\1")

splicemode_colnames <- rep("splicemode", 6)

column_of_splicemode_names_appendedwithsplicemode <- purrr::map2(.x = column_of_summary_tables_per_splicemode, .y = column_of_splicemode_names, ~cbind(.x, splicemode=.y))

```

### combining the summary tables for each splice mode into a large table

```{r}



```

